### ISSUE TITLE
Add Istiod internal caching for generated XDS responses

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#26732]
* EDS: #26024
* CDS: https://github.com/istio/istio/pull/33338
* LDS: Not started
* RDS: https://github.com/istio/istio/pull/34243
* SDS: done in initial PR

**Comment by rshriram** [comment]
For posterity, can we have a rough write up of how the caching will be done for each XDS resource, in this issue? It will help spot design issues of any, as well as discuss caveats for various conditions like same service entry in different namespaces, etc.

**Comment by rshriram** [comment]
With regard to caching, I think we need to determine a few things before hand:
1. Unit of caching - individual resource or the entire list [pros of individual resource is that memory consumption may be low but the code will become complicated as we will have to apply transformations before sending (envoyFilters, possibly sidecars, etc.). Caching an entire list means we will have many sets of caches for different "groups" of proxies but we wont have to do any transformation before sending the data out]
2. Cache keys - determine how we are going to reuse a cached entry. 
    Istio Configs for a proxy are selected based on its labels, its namespace and whatever is visible to that namespace (exportTo).
    Generated configs for a proxy depend on the proxy version, proxy type, proxy domain (which is used to auto fill vs/dr), one or more metadata embedded in the proxy, as well as the Istio configs for that proxy.
  So its safe to say that a key containing a redacted version of the XDS Node metadata (including namespace, labels, istio_meta stuff, but excluding the proxy ID, IP, or anything else that is system generated) can be used as a cache key. But if we do this, we automatically fall into the "caching entire list" model because this key will be the same for all pods of a deployment [i.e. all pods sharing the exact same set of labels in a given namespace].

3. Invalidation. As @ramaraochavali mentioned in the EDS PR, we may either need some LRU or a periodic cache refresh where we throw away all the existing data and regenerate the cache every 30s or so (this will result in a saw toothed memory consumption curve at worst, but memory wont grow unbounded). I think the periodic refresh is an easier way to lock down memory leaks.

4. Correctness proof. The cache we have is not the same as some DB lookup cache as this is generated data that takes into account a complex set of attributes from the environment. We need some way (even if its on paper design) to run through the correctness of a cached entry - i.e. is it truly applicable to the given proxy or are we sending it the wrong set of configs. For example, in the EDS cache implementation we have today, the key does not include proxy Labels. So, if two proxies have two different Sidecar crds, where each proxy imports a service entry with same host, but containing different sets of endpoints, the cache would return an incorrect data. [You pointed this issue correctly in our previous discussion but I failed to see the nuances].

If we were to work backwards from (4), what is the easiest way to construct the cache (at cost of memory?) with provably correct response data? I think that caching at fine grained levels and then selectively applying EnvoyFilters, Sidecars, or what not will not be an easy thing to prove for sidecar proxies [the case I provided above is an example]. But I am interested in hearing counter arguments to this assumption.
OTOH, caching at coarse granularities (like using the node metadata with wildcards for certain fields) makes it easy to reason about the correctness albeit at higher memory cost. We will end up having N sets of complete XDS responses in memory where N is the number of deployments in the system - this will quickly result in OOM unless we aggressively evict items.

[FTR, this is also part of the reason why I am not convinced that an intermediate XDS relay will be of any use as there is too much logic involved in determining whether or not something is resuable].

Thoughts @howardjohn / @hzxuzhonghu / @ramaraochavali ?

**Comment by rshriram** [comment]
Update: 
 Here is an alternative cache key for better cache utilization.

 We can have a two level cache hierarchy. Given a proxy, we can compute set of all configs for that proxy (including networking, security, etc.) and then compute some hash of all these configs. This will be the first level cache key. Within each level, we can cache outbound XDS (rds/cds/lds/eds) in a second level hierarchy, keyed by the node metadata (that includes locality, cluster, network, ISTIO_META vars). We can then always compute the inbound as we are doing today.

So, if two proxies in the same namespace have two different Sidecars or EnvoyFilters mutating different things for same EDS/CDS/LDS/RDS target, then the cache key will naturally be different. But within this set, we should be able to reuse the data across all proxies. If we go one step further to keep the EDS/CDS cache different from LDS/RDS, then LDS/RDS can be used universally based on ISTIO_METAs but irrespective of the proxy's locality/network/cluster while EDS/CDS can be further refined based on both ISTIO_METAs and the locality/network/cluster.

**Comment by howardjohn** [comment]
> 1 Unit of caching 

In my opinion, per resource is the only viable way. If you look at resources across a cluster, there is a *huge* amount of overlap. If you look at resource collections, there is a very small amount - likely just proxies in the same exact deployment in many cases. As a result, our hit rate is way lower, and we need to store way more redundant config (== OOM). With collection storage we have potential N^2 storage - with per resource storage is likely actually smaller memory usage, since instead of generating each resource for each proxy, we generate each resource 1x and send it to each proxy.

If we ignore other proxies, its still be much better. For example, with EDS, with collection caching, if we change a Service, we now invalidate the entire collection. With resource caching, we just invalidate the changed resource - we can still keep all the other resources cached.

> 2 Cache keys

I do not think we should be caching based on the proxy identity. Rather, we should be caching based on the inputs to the resources we are generating. For example, a given CLA response is based almost entirely on the Service/Endpoints and the network/cluster of a proxy, but also DR for locality lb. The cache need not be aware of any impacts of Sidecar scoping -- instead, the caching logic takes in the inputs needed to construct the resource and uses them as keys. When these change, we invalidate the cache. This means, for example, that if a Sidecar resource no longer selects the DR that applies some locality LB config, we will now have a new key (since DR is part of the key), and get the correct response.

Trying to do things at a higher level is too complicated IMO. There are many many factors from the proxy that determine the output; it is much simpler to look closer to the resource generation to identify what are the true input keys.

In order to ensure we do not take dependencies on resources not part of the caching key/logic, we need to make the inputs to the config gen as narrow as possible. For example, instead of passing mega-structs like PluginParams which attempt to take essentially every bit of information we have about the state of the cluster, we instead finely scope down the inputs. A specific example would be passing in the DestinationRule that will actually be applied, which is read from the PushContext before looking into the cache, rather than PushContext and doing the lookup after the cache lookup.

> 3 Invalidation

I agree we need some periodic cleanup to prevent cases where keys will never be used again. If a DR is removed, those keys will be removed, but there are ways for things to implicitly never be used which we do not invalidate (new Sidecar means nothing uses the DR, XDS client goes away, etc). For these cases, some LRU type system makes sense. Dropping the entire cache every 30s seems far to aggressive IMO. I doubt we will see sawtooth memory, and instead see needless CPU spikes - but we will need to validate in large scale clusters.

It should be noted that the EDS cache had a trivial impact on memory in large clusters during my testing.

> 4 Correctness proof

Agreed we need to prove correctness, made simpler by the pattern I mentioned in cache keys.

> So, if two proxies have two different Sidecar crds, where each proxy imports a service entry with same host, but containing different sets of endpoints, the cache would return an incorrect data.

The code handles this just fine, we key on the hostname.namespace, which is unique across the cluster. This is exactly why we key on the inputs to the config gen rather than the proxy metadata - so we are keying on *exactly* what is used by the cache, rather than attempting to re-implement the logic at a higher level based on some complex heuristics.

---

In my opinion, the EDS cache implementation makes the right choices. It is simple to prove correctness (just look at EndpointBuilder and usages), extremely high cache hit rate (even in our integration tests, which apply a bunch of unique config in a short period of time see a 98% hit rate when I looked), and simple implementation. I would encourage us to use this model for the other resources.

I have been focusing on CDS next as I find it is the next most resource intensive and simplest to cache (followed by RDS, then LDS - LDS will be very challenging). The main issue with CDS is the resources are currently generated by the "monolithic" `BuildClusters` which builds all of the clusters needed, whereas EDS we have `loadAssignmentsForCluster` which generates a single CLA. This choice is not a requirement, and its actually fairly simple to change provided we refactor how we apply EnvoyFilters: https://github.com/istio/istio/pull/26734.

Once that is done, we then can create a similar ClusterBuilder (we already have this actually, so just refactor it) mirroring EndpointBuilder. In this, we will declare all the dependencies to generate a cluster. I am still exploring this, but right now I think its similar to:
* destinationRule
* direction
* networkView
* proxyType
* service
* service port

Its possible I missed some, I will be taking a closer look soon as I work on the implementation. The main takeaway is that the approach used by EDS (which IMO is working quite well) will be extremely similar to CDS - the cache keys are nearly identical even. For RDS and LDS we will need to evaluate, which I have not yet started.

**Comment by hzxuzhonghu** [comment]
>>So, if two proxies have two different Sidecar crds, where each proxy imports a service entry with same host, but containing different sets of endpoints, the cache would return an incorrect data.

>The code handles this just fine, we key on the hostname.namespace, which is unique across the cluster. This is exactly why we key on the inputs to the config gen rather than the proxy metadata - so we are keying on exactly what is used by the cache, rather than attempting to re-implement the logic at a higher level based on some complex heuristics.

But it may override the value incorrectly, there is a probability, but i don't think a user will use ServiceEntry like that.

**Comment by hzxuzhonghu** [comment]
I would agree with the per resource cache. And furthermore also consider cache for cluster scope envoyfilters, because the patch is not very efficient as i know.


For the listener, it is the most complicated, as it is built in many stages, but the work complexity here is in proportion to the cache granularity. For example, we can cache only for the outbound direction, and if necessary we can split the lds generating process

In general, I propose each one take one kind of xds to investigate and so that we can make sure all xds resources can be done in 1.8.
release.

**Comment by howardjohn** [comment]
I think rds and lds will be really hard since they are composed of multiple resources (many virtual service and service) whereas clusters and eds are all made up of a single resource (one dr applied to one service) so we need to figure out how we will handle that. cds should be no problem for 1.8 though

**Comment by ramaraochavali** [comment]
+1 for per resource cache. Aggregation might not give the benefits vs. memory cost. 

And we start using lruCache and see if we can gets some stats out of it - measure cache efficiency as we roll out to other resources and watch in different setups.

> And furthermore also consider cache for cluster scope envoyfilters, because the patch is not very efficient as i know.

@hzxuzhonghu What do you mean by above? I thought in CDS cache we store patched cluster only?

**Comment by hzxuzhonghu** [comment]
I mean some envoyfilter may only be applied to some specific proxies.

**Comment by rshriram** [comment]
With regard to the cache keys, think about an upgrade scenario.. If in 1.9, we decide to introduce a new functionality based on some envoy changes, we will most likely have something like `if proxyVersion >= and nodeMetadata[blah]=true`. If we build the cache without the proxy metadata (which includes version, among other things), we will send the incorrect config to a 1.8 proxy causing a crash. Rather than continue to add more and more things to the cache key, its easier to have parts of the node metadata (not the entire proxy identity) to the cache key to avoid this scenario.

Now, with regard to per-resource vs collection, what is the goal of the caching work? To reduce compute or memory usage or both? If we are computing all the inputs required to generate a particular micro config like CLA or cluster, then we have already done a good chunk of work haven't we? like selecting the destination rules for a proxy, the sidecars, iterating through all the 1000s of services, and so on.. At this point, the cpu saving is in the work done by that cluster generation code. This is the reason I was suggesting for collection caching at coarser granularities.. 

 The CDS code goes like this

```
  1. Compute sidecar CRD for a proxy (which in turn computes list of all services, dest rules, virtual services, etc. for proxy)
  2. for each service visible to the proxy (computed from Sidecar crd)
       2.1 getDestRule for service for proxy
       2.2 generateOutboundCluster based on service, proxyMetadata, destRule
  3. applyEnvoyFilterChanges on collection
  4. Generate inbound clusters based on proxy inbound ports, and the Sidecar crd's ingressListeners
```

And what you are suggesting is 

```
   1. Compute sidecar CRD for a proxy (which in turn computes list of all services, dest rules, virtual services, etc. for proxy)
   2. for each service visible to the proxy (computed from Sidecar crd)
       2.1 getDestRule & EnvoyFilters for service for proxy
       2.2 generateCluster based on service, destRule
       2.3 applyEnvoyFilterChanges on the individual cluster
   3. applyEnvoyFilter Adds
   4. Generate inbound clusters based on proxy inbound ports, and the Sidecar crd's ingressListeners
```

For starters, there is correctness issue in my naive algo already as a merge could refer to a cluster added by another EnvoyFilter. Same goes for replace. I assume you have taken this into account, but you are still not going to be able to handle the proxyMetadata thing (we generated SDS configs for a long time based on proxy metadata). Assuming we handle all this, I think the memory gain is clear but the compute gain is not unless the bulk of compute is going towards marshalling to any.Any (which we are caching at fine grained levels). If the max gains are from avoiding the marshalling to any.Any, then I agree with all of your assessments above, but we still need to tweak the cache keys.


But do think about this

What if we have a per Sidecar crd (all except root namespace sidecar crd, which we can copy logically to each namespace) cache? this should already include root namespace EnvoyFilter and namespace global EnvoyFilter as @hzxuzhonghu suggests, and should have automatically taken into account the destination rule caveats. We should be careful to treat the no Sidecar crd case properly. 

With this, whenever a proxy comes in, we can identify the Sidecar crd to which it belongs to (most of time, its namespace global crd or root namespace crd) [this cache can still be LRUed]. Once we know the sidecar crd, we know the cds/rds/lds for the proxy (if the proxy has a workload specific crd, then its likely to not be cached). We can iterate through this collection, take the array entries that are not modified as is, and these that are modified by workload specific EnvoyFilters can be copy-on-write. The cached set of clusters will not include the ones that are based on locality (we can identify these easily).

**Comment by howardjohn** [comment]
> Rather than continue to add more and more things to the cache key, its easier to have parts of the node metadata (not the entire proxy identity) to the cache key to avoid this scenario.

Good point. For EDS we actually do not pass any of the model.Proxy around. For the other ones which are more proxy dependent we may need some other solution here

> Now, with regard to per-resource vs collection, what is the goal of the caching work? To reduce compute or memory usage or both?

Primarily compute, memory is secondary

> If we are computing all the inputs required to generate a particular micro config

This part is trivial. It represents ~1% of the CPU used by the system today. Almost all of processing time is spent on the XDS config gen (and GC associated with that)

> For starters, there is correctness issue in my naive algo already as a merge could refer to a cluster added by another EnvoyFilter. 

Merge -> add -> remove is the ordering: https://github.com/istio/istio/blob/07782fd8d2ff0225af494b172dc459e5c9ace04a/pilot/pkg/networking/core/v1alpha3/envoyfilter/cluster_patch.go#L69

> but you are still not going to be able to handle the proxyMetadata thing

We can just pass in sds enabled as part of the ClusterBuilder rather than passing the entire model.Proxy. This explicitly declares the dependencies we can use for the cluster gen. (just an example - sds is always on now)


> If the max gains are from avoiding the marshalling to any.Any, then I agree with all of your assessments above, but we still need to tweak the cache keys.

The vast majority of the CPU is spent on any.Any marshalling

--

Its hard to hypothesize about this, it may make sense to just try out some prototypes of variants and compare them directly.

**Comment by howardjohn** [comment]
This is done. We can do more of course, but the core is there

**Comment by shonecyx** [comment]
@hzxuzhonghu  what's the plan for LDS cache? Is it done or still not started?

**Comment by hzxuzhonghu** [comment]
LDS is quite different, one listener could be shared by mant service, the factors (services, vs, sidecars, etc) number that can effect on it could be huge.

**Comment by shonecyx** [comment]
> LDS is quite different, one listener could be shared by mant service, the factors (services, vs, sidecars, etc) number that can effect on it could be huge.

@hzxuzhonghu  In our case, since we enable gateway bind by default, instead of sharing the any_address listener with a lot of filterChains, envoy will listen on the VIPs and there are a bunch of listeners(each only has one filter chain) for those VIPs. It looks to me that we can do some optimization for this case by the LDS cache.
